File: src/main/java/space/accident/api/interfaces/tileentity/redstone/IRedStoneEmitter.java
package space.accident.api.interfaces.tileentity.redstone;

import space.accident.api.interfaces.tileentity.IWorldInteraction;

/**
 * This File has just internal Information about the RedStone State of a TileEntity
 */
public interface IRedStoneEmitter extends IWorldInteraction {
	/**
	 * gets the RedStone Level the TileEntity should emit to the given Output Side
	 */
	int getOutputRedStoneSignal(int side);
	
	/**
	 * sets the RedStone Level the TileEntity should emit to the given Output Side
	 * <p/>
	 * Do not use this if ICoverable is implemented. ICoverable has @getInternalOutputRedstoneSignal for Machine internal Output RedStone, so that it doesnt conflict with Cover RedStone.
	 * This sets the true RedStone Output Signal. Only Cover Behaviors should use it, not MetaTileEntities.
	 */
	void setOutputRedStoneSignal(int side, int aStrength);
	
	/**
	 * gets the RedStone Level the TileEntity should emit to the given Output Side
	 */
	int getStrongOutputRedStoneSignal(int side);
	
	/**
	 * sets the RedStone Level the TileEntity should emit to the given Output Side
	 * <p/>
	 * Do not use this if ICoverable is implemented. ICoverable has @getInternalOutputRedStoneSignal for Machine internal Output RedStone, so that it doesnt conflict with Cover RedStone.
	 * This sets the true RedStone Output Signal. Only Cover Behaviors should use it, not MetaTileEntities.
	 */
	void setStrongOutputRedStoneSignal(int side, int aStrength);
	
	/**
	 * Gets the Output for the comparator on the given Side
	 */
	int getComparatorValue(int side);
	
	/**
	 * Get the RedStone output signal strength for a given side
	 */
	default int getSideRedStone(int side) {
		return 0;
	}
}

File: src/main/kotlin/space/accident/api/interfaces/te/IRedstoneEmitter.kt
package space.accident.api.interfaces.te


/**
 * This File has just internal Information about the Redstone State of a TileEntity
 */
interface IRedstoneEmitter : IHasWorldObjectAndCoords {
    /**
     * gets the Redstone Level the TileEntity should emit to the given Output Side
     */
    fun getOutputRedstoneSignal(aSide: Byte): Byte

    /**
     * sets the Redstone Level the TileEntity should emit to the given Output Side
     *
     *
     * Do not use this if ICoverable is implemented. ICoverable has @getInternalOutputRedstoneSignal for Machine internal Output Redstone, so that it doesnt conflict with Cover Redstone.
     * This sets the true Redstone Output Signal. Only Cover Behaviors should use it, not MetaTileEntities.
     */
    fun setOutputRedstoneSignal(aSide: Byte, aStrength: Byte)

    /**
     * gets the Redstone Level the TileEntity should emit to the given Output Side
     */
    fun getStrongOutputRedstoneSignal(aSide: Byte): Byte

    /**
     * sets the Redstone Level the TileEntity should emit to the given Output Side
     *
     *
     * Do not use this if ICoverable is implemented. ICoverable has @getInternalOutputRedstoneSignal for Machine internal Output Redstone, so that it doesnt conflict with Cover Redstone.
     * This sets the true Redstone Output Signal. Only Cover Behaviors should use it, not MetaTileEntities.
     */
    fun setStrongOutputRedstoneSignal(aSide: Byte, aStrength: Byte)

    /**
     * Gets the Output for the comparator on the given Side
     */
    fun getComparatorValue(aSide: Byte): Byte

    /**
     * Get the redstone output signal strength for a given side
     */
    fun getGeneralRS(aSide: Byte): Byte {
        return 0
    }
}
